File: sdk/typescript/src/pod.ts
import { execSync, execFileSync } from 'child_process';
import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';
import { ensureInstalled } from './installer';

export interface PodOptions {
  config?: string;
  preset?: string;
  mode?: 'standard' | 'full';
  mountCwd?: boolean;
}

export interface RunOptions {
  root?: boolean;
  env?: Record<string, string>;
  capture?: boolean;
  display?: boolean;
  audio?: boolean;
  background?: boolean;
}

export interface CommitOptions {
  exclude?: string[];
  output?: string;
  rollbackRest?: boolean;
}

export interface ResizeOptions {
  cpus?: number;
  memory?: string;
  tmpSize?: string;
  maxPids?: number;
  gpu?: boolean;
}

export class Pod {
  readonly name: string;
  private config?: string;
  private preset?: string;
  private mode: string;
  private binary: string;

  private mountCwd: boolean;

  private constructor(name: string, opts: PodOptions = {}) {
    this.name = name;
    this.config = opts.config;
    this.preset = opts.preset;
    this.mode = opts.mode || getMode();
    this.mountCwd = opts.mountCwd !== false; // default true
    this.binary = ensureInstalled();
  }

  /**
   * Create a new pod with setup.
   */
  static async create(name: string, opts: PodOptions = {}): Promise<Pod> {
    const pod = new Pod(name, opts);
    if (!pod.exists()) {
      pod.init();
    }
    return pod;
  }

  /**
   * Create a pod, run a callback, then destroy + gc.
   * Node.js equivalent of Python's context manager.
   *
   * @example
   * await Pod.with('my-agent', { config: 'pod.yaml' }, async (pod) => {
   *   pod.run('python3 agent.py');
   *   pod.commit(['src/'], { rollbackRest: true });
   * });
   */
  static async with(name: string, opts: PodOptions, fn: (pod: Pod) => void | Promise<void>): Promise<void> {
    const pod = await Pod.create(name, opts);
    try {
      await fn(pod);
    } finally {
      pod.destroy();
      Pod.gc();
    }
  }

  /**
   * Wrap an existing pod (no init).
   */
  static wrap(name: string, opts: PodOptions = {}): Pod {
    return new Pod(name, opts);
  }

  /**
   * Initialize the pod (create + setup).
   */
  init(opts?: { verbose?: boolean }): void {
    const args = ['init', this.name];
    if (this.config) args.push('-c', this.config);
    else if (this.preset) args.push('--preset', this.preset);
    if (opts?.verbose) args.push('--verbose');
    this.exec(args);
  }

  /**
   * Run a command inside the pod.
   */
  run(command: string, opts: RunOptions = {}): string | void {
    const args = ['run', this.name];
    if (opts.root) args.push('--root');
    if (this.mountCwd) args.push('-w');
    if (opts.display) args.push('-d');
    if (opts.audio) args.push('-a');
    if (opts.background) args.push('-b');
    if (opts.env) {
      for (const [k, v] of Object.entries(opts.env)) {
        args.push('--env', `${k}=${v}`);
      }
    }
    args.push('--', 'sh', '-c', command);
    if (opts.capture) {
      return this.exec(args, true);
    }
    this.exec(args);
  }

  /**
   * Run inline code inside the pod.
   * Writes to a temp file in the overlay, executes, cleans up.
   */
  runScript(code: string, opts: RunOptions & { interpreter?: string } = {}): string | void {
    const interp = opts.interpreter || 'python3';
    const encoded = Buffer.from(code).toString('base64');
    const cmd = `echo "${encoded}" | base64 -d > /tmp/.envpod-script && ${interp} /tmp/.envpod-script && rm -f /tmp/.envpod-script`;
    return this.run(cmd, opts);
  }

  /**
   * Copy a local file into the pod and run it.
   * Auto-detects interpreter from file extension.
   */
  runFile(path: string, opts: RunOptions & { interpreter?: string } = {}): string | void {
    const { readFileSync } = require('fs');
    const code = readFileSync(path, 'utf-8');
    const ext = path.split('.').pop()?.toLowerCase();
    const interp = opts.interpreter || ({
      py: 'python3', js: 'node', ts: 'npx tsx',
      sh: 'bash', rb: 'ruby', pl: 'perl',
    } as Record<string, string>)[ext || ''] || 'bash';
    return this.runScript(code, { ...opts, interpreter: interp });
  }

  /**
   * Mount a host directory into the pod (COW isolated).
   */
  mount(path: string, readonly: boolean = true): void {
    const args = ['mount', this.name, path];
    if (readonly) args.push('--readonly');
    this.exec(args);
  }

  /**
   * Copy a local file into the pod's overlay.
   */
  inject(localPath: string, podPath: string = '/tmp/', executable: boolean = false): void {
    const { readFileSync } = require('fs');
    const { basename } = require('path');
    const content = readFileSync(localPath);
    const encoded = content.toString('base64');
    const name = basename(localPath);
    const dest = podPath.endsWith('/') ? `${podPath}${name}` : podPath;
    let cmd = `echo "${encoded}" | base64 -d > ${dest}`;
    if (executable) cmd += ` && chmod +x ${dest}`;
    this.run(cmd, { root: true });
  }

  /**
   * Show filesystem changes.
   */
  diff(opts?: { all?: boolean; json?: boolean }): string {
    const args = ['diff', this.name];
    if (opts?.all) args.push('--all');
    if (opts?.json) args.push('--json');
    return this.exec(args, true);
  }

  /**
   * Commit overlay changes to host.
   */
  commit(paths: string[] = [], opts: CommitOptions = {}): void {
    const args = ['commit', this.name];
    if (opts.rollbackRest) args.push('--rollback-rest');
    if (opts.output) args.push('--output', opts.output);
    if (opts.exclude) {
      for (const e of opts.exclude) args.push('--exclude', e);
    }
    args.push(...paths);
    this.exec(args);
  }

  /**
   * Discard all overlay changes.
   */
  rollback(): void {
    this.exec(['rollback', this.name]);
  }

  /**
   * Show audit log or security analysis.
   */
  audit(opts?: { security?: boolean; json?: boolean }): string {
    const args = ['audit', this.name];
    if (opts?.security) args.push('--security');
    if (opts?.json) args.push('--json');
    return this.exec(args, true);
  }

  /**
   * Show pod status.
   */
  status(): string {
    return this.exec(['status', this.name], true);
  }

  /**
   * Start pod in background.
   */
  start(): void {
    this.exec(['start', this.name]);
  }

  /**
   * Stop pod.
   */
  stop(): void {
    this.exec(['stop', this.name]);
  }

  /**
   * Freeze pod.
   */
  lock(): void {
    this.exec(['lock', this.name]);
  }

  /**
   * Resume frozen pod.
   */
  unlock(): void {
    this.exec(['unlock', this.name]);
  }

  /**
   * Restart the pod (stop + start).
   */
  restart(): void {
    this.exec(['restart', this.name]);
  }

  /**
   * Terminate pod processes and rollback changes.
   */
  kill(): void {
    this.exec(['kill', this.name]);
  }

  /**
   * Remove pod entirely.
   */
  destroy(): void {
    try {
      this.exec(['destroy', this.name]);
    } catch {
      // Pod may already be destroyed
    }
  }

  /**
   * Resize pod resources.
   */
  resize(opts: ResizeOptions): void {
    const args = ['resize', this.name];
    if (opts.cpus !== undefined) args.push('--cpus', String(opts.cpus));
    if (opts.memory) args.push('--memory', opts.memory);
    if (opts.tmpSize) args.push('--tmp-size', opts.tmpSize);
    if (opts.maxPids !== undefined) args.push('--max-pids', String(opts.maxPids));
    if (opts.gpu !== undefined) args.push('--gpu', String(opts.gpu));
    this.exec(args);
  }

  /**
   * Store a secret in the vault.
   */
  vaultSet(key: string, value: string): void {
    this.exec(['vault', this.name, 'set', key, value]);
  }

  /**
   * Create a snapshot of the pod's current overlay state.
   */
  snapshotCreate(name?: string): void {
    const args = ['snapshot', this.name, 'create'];
    if (name) args.push('--name', name);
    this.exec(args);
  }

  /**
   * Restore a snapshot.
   */
  snapshotRestore(name: string): void {
    this.exec(['snapshot', this.name, 'restore', name]);
  }

  /**
   * List all snapshots.
   */
  snapshotList(): string {
    return this.exec(['snapshot', this.name, 'ls'], true);
  }

  /**
   * Delete a snapshot.
   */
  snapshotDestroy(name: string): void {
    this.exec(['snapshot', this.name, 'destroy', name]);
  }

  /**
   * Show pod output logs.
   */
  logs(): string {
    return this.exec(['logs', this.name], true);
  }

  /**
   * Get pod info (name, status, IP, display URL, etc.).
   */
  info(): Record<string, any> {
    try {
      const output = this.exec(['ls', '--json'], true);
      const pods = JSON.parse(output) as Array<Record<string, any>>;
      return pods.find(p => p.name === this.name) || {};
    } catch {
      return {};
    }
  }

  /**
   * Get the noVNC display URL if web display is enabled.
   */
  get displayUrl(): string | null {
    const info = this.info();
    return info.display_url || info.novnc_url || null;
  }

  /**
   * Get the pod's IP address.
   */
  get ip(): string | null {
    const info = this.info();
    return info.ip || info.pod_ip || null;
  }

  /**
   * Create pod and save as base for fast cloning.
   */
  initWithBase(opts?: { verbose?: boolean; baseName?: string }): void {
    const args = ['init', this.name];
    if (opts?.baseName) {
      args.push(`--create-base=${opts.baseName}`);
    } else {
      args.push('--create-base');
    }
    if (this.config) args.push('-c', this.config);
    else if (this.preset) args.push('--preset', this.preset);
    if (opts?.verbose) args.push('--verbose');
    this.exec(args);
  }

  /**
   * Clone a pod from a base (fast — ~8ms).
   */
  static clone(source: string, name: string, opts: PodOptions = {}): Pod {
    const pod = new Pod(name, opts);
    pod.exec(['clone', source, name]);
    return pod;
  }

  /**
   * Clone from base, run command, optionally save results, destroy.
   * The fastest governed execution — ~8ms clone, run, commit, destroy.
   * Equivalent to `docker run --rm` but with governance.
   */
  static disposable(base: string, name: string, command: string, opts: {
    commitPaths?: string[];
    output?: string;
    mode?: 'standard' | 'full';
    root?: boolean;
    env?: Record<string, string>;
  } = {}): string | null {
    const pod = Pod.clone(base, name, { mode: opts.mode });
    try {
      pod.run(command, { root: opts.root, env: opts.env });
      if (opts.commitPaths) {
        const diff = pod.diff();
        pod.commit(opts.commitPaths, { output: opts.output, rollbackRest: true });
        return diff;
      }
      return null;
    } finally {
      pod.destroy();
    }
  }

  /**
   * Clean up orphaned resources (iptables, cgroups, netns).
   */
  static gc(): void {
    const binary = ensureInstalled();
    try {
      execFileSync('sudo', [binary, 'gc'], { stdio: 'pipe' });
    } catch {}
  }

  /**
   * Check if pod exists.
   */
  exists(): boolean {
    try {
      const output = this.exec(['ls', '--json'], true);
      const pods = JSON.parse(output) as Array<{ name: string }>;
      return pods.some(p => p.name === this.name);
    } catch {
      return false;
    }
  }

  private exec(args: string[], capture?: true): string;
  private exec(args: string[], capture?: false): void;
  private exec(args: string[], capture: boolean = false): string | void {
    const cmd = this.mode === 'full'
      ? ['sudo', this.binary, ...args]
      : [this.binary, ...args];

    if (capture) {
      return execFileSync(cmd[0], cmd.slice(1), {
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'pipe'],
      });
    }
    execFileSync(cmd[0], cmd.slice(1), {
      stdio: 'inherit',
    });
  }
}

function getMode(): string {
  // Check environment variable
  const envMode = process.env.ENVPOD_MODE;
  if (envMode === 'standard' || envMode === 'full') return envMode;

  // Check saved config
  const configDir = join(homedir(), '.config', 'envpod');
  const configFile = join(configDir, 'sdk.json');
  if (existsSync(configFile)) {
    try {
      const config = JSON.parse(readFileSync(configFile, 'utf-8'));
      if (config.mode) return config.mode;
    } catch {}
  }

  // Interactive prompt
  const readline = require('readline');
  const rl = readline.createInterface({ input: process.stdin, output: process.stderr });

  console.error('\n  envpod — choose isolation mode:');
  console.error('    [1] Standard — full governance, no sudo needed');
  console.error('        (no cgroup limits, no network namespace)');
  console.error('    [2] Full — complete isolation + governance (requires sudo)');
  console.error('        (cgroup limits, network namespace, DNS filtering)\n');

  // Synchronous prompt via spawnSync
  const result = require('child_process').spawnSync('bash', ['-c', 'read -p "  Choice [1/2]: " c; echo $c'], {
    stdio: ['inherit', 'pipe', 'inherit'],
    encoding: 'utf-8',
  });
  const choice = (result.stdout || '').trim();
  rl.close();

  const mode = choice === '1' ? 'standard' : 'full';

  // Save preference
  try {
    mkdirSync(configDir, { recursive: true });
    writeFileSync(configFile, JSON.stringify({ mode }));
    console.error(`  Saved to ${configFile}`);
  } catch {}

  return mode;
}

File: sdk/python/envpod/pod.py
"""Pod — the core abstraction for governed AI agent environments."""

import json
import subprocess
import shutil
from typing import List, Optional


class PodError(Exception):
    """Raised when an envpod command fails."""
    pass


class Pod:
    """A governed environment for an AI agent.

    Args:
        name: Pod name (must be unique).
        config: Path to pod.yaml config file (for init).
        preset: Built-in preset name (alternative to config).
        mode: "standard" (no sudo) or "full" (sudo, complete isolation).
              If None, prompts interactively on first use.
    """

    def __init__(self, name: str, config: Optional[str] = None,
                 preset: Optional[str] = None, mode: Optional[str] = None,
                 persist: bool = False):
        self.name = name
        self._config = config
        self._preset = preset
        self._mode = mode or _get_mode()
        self._initialized = False
        self._persist = persist

        # Ensure envpod binary is available
        from envpod.installer import ensure_installed
        ensure_installed()

    def __enter__(self):
        if not self.exists():
            self.init(config=self._config, preset=self._preset)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        if not self._persist:
            self.destroy()
            self.gc()
        return False

    def init(self, config: Optional[str] = None, preset: Optional[str] = None,
             verbose: bool = False, mount_cwd: bool = True) -> None:
        """Create and set up the pod.

        Args:
            config: Path to pod.yaml config file.
            preset: Built-in preset name.
            verbose: Show live setup output.
            mount_cwd: Mount current working directory into the pod (default True).
        """
        args = ["init", self.name]
        cfg = config or self._config
        pre = preset or self._preset
        if cfg:
            args.extend(["-c", cfg])
        elif pre:
            args.extend(["--preset", pre])
        if verbose:
            args.append("--verbose")
        self._run(args)
        self._initialized = True
        self._mount_cwd = mount_cwd

    def run(self, command: str, root: bool = False,
            env: Optional[dict] = None, capture: bool = False,
            display: bool = False, audio: bool = False,
            background: bool = False) -> Optional[str]:
        """Run a command inside the pod.

        Args:
            command: Shell command to execute.
            root: Run as root inside the pod.
            env: Additional environment variables.
            capture: If True, capture and return stdout instead of printing.
            display: Enable display forwarding (Wayland/X11).
            audio: Enable audio forwarding (PipeWire/PulseAudio).
            background: Run in background (detached).

        Returns:
            stdout as string if capture=True, else None.
        """
        args = ["run", self.name]
        if root:
            args.append("--root")
        if getattr(self, '_mount_cwd', True):
            args.append("-w")
        if display:
            args.append("-d")
        if audio:
            args.append("-a")
        if background:
            args.append("-b")
        if env:
            for k, v in env.items():
                args.extend(["--env", f"{k}={v}"])
        args.append("--")
        args.extend(["sh", "-c", command])
        return self._run(args, capture=capture)

    def run_script(self, code: str, interpreter: str = "python3",
                   root: bool = False, env: Optional[dict] = None,
                   capture: bool = False) -> Optional[str]:
        """Run inline code inside the pod.

        Writes code to a temp file in the overlay, executes it, cleans up.

        Args:
            code: Source code string to execute.
            interpreter: Interpreter to use (python3, node, bash, etc.)
            root: Run as root inside the pod.
            env: Additional environment variables.
            capture: If True, capture and return stdout.
        """
        import base64
        encoded = base64.b64encode(code.encode()).decode()
        shell_cmd = (
            f'echo "{encoded}" | base64 -d > /tmp/.envpod-script && '
            f'{interpreter} /tmp/.envpod-script && '
            f'rm -f /tmp/.envpod-script'
        )
        return self.run(shell_cmd, root=root, env=env, capture=capture)

    def run_file(self, path: str, interpreter: Optional[str] = None,
                 root: bool = False, env: Optional[dict] = None,
                 capture: bool = False) -> Optional[str]:
        """Copy a local file into the pod and run it.

        Args:
            path: Path to local file.
            interpreter: Interpreter (auto-detected from extension if not set).
            root: Run as root inside the pod.
            env: Additional environment variables.
            capture: If True, capture and return stdout.
        """
        import os
        if not os.path.exists(path):
            raise PodError(f"file not found: {path}")

        with open(path) as f:
            code = f.read()

        if interpreter is None:
            ext = os.path.splitext(path)[1].lower()
            interpreter = {
                '.py': 'python3',
                '.js': 'node',
                '.ts': 'npx tsx',
                '.sh': 'bash',
                '.rb': 'ruby',
                '.pl': 'perl',
            }.get(ext, 'bash')

        return self.run_script(code, interpreter=interpreter, root=root,
                               env=env, capture=capture)

    def mount(self, path: str, readonly: bool = True) -> None:
        """Mount a host directory into the pod (COW isolated).

        Args:
            path: Host path to mount (also the mount point inside the pod).
            readonly: If True, mount as read-only.
        """
        args = ["mount", self.name, path]
        if readonly:
            args.append("--readonly")
        self._run(args)

    def inject(self, local_path: str, pod_path: str = "/tmp/",
               executable: bool = False) -> None:
        """Copy a local file into the pod's overlay.

        Args:
            local_path: Path to file on host.
            pod_path: Destination path inside pod (directory or full path).
            executable: If True, chmod +x after copying.
        """
        import os
        import base64

        if not os.path.exists(local_path):
            raise PodError(f"file not found: {local_path}")

        with open(local_path, "rb") as f:
            encoded = base64.b64encode(f.read()).decode()

        basename = os.path.basename(local_path)
        if pod_path.endswith("/"):
            dest = f"{pod_path}{basename}"
        else:
            dest = pod_path

        cmd = f'echo "{encoded}" | base64 -d > {dest}'
        if executable:
            cmd += f' && chmod +x {dest}'

        self.run(cmd, root=True)

    def diff(self, all_changes: bool = False, json_output: bool = False) -> str:
        """Show filesystem changes in the pod's overlay.

        Returns:
            Diff output as string.
        """
        args = ["diff", self.name]
        if all_changes:
            args.append("--all")
        if json_output:
            args.append("--json")
        return self._run(args, capture=True)

    def commit(self, *paths: str, exclude: Optional[List[str]] = None,
               output: Optional[str] = None, rollback_rest: bool = False) -> None:
        """Commit overlay changes to the host filesystem.

        Args:
            paths: Specific paths to commit (commits all if empty).
            exclude: Paths to exclude from commit.
            output: Export to this directory instead of host filesystem.
            rollback_rest: Rollback all uncommitted changes after commit.
        """
        args = ["commit", self.name]
        if rollback_rest:
            args.append("--rollback-rest")
        if output:
            args.extend(["--output", output])
        if exclude:
            for e in exclude:
                args.extend(["--exclude", e])
        args.extend(paths)
        self._run(args)

    def rollback(self) -> None:
        """Discard all overlay changes."""
        self._run(["rollback", self.name])

    def audit(self, security: bool = False, json_output: bool = False) -> str:
        """Show audit log or run security analysis.

        Returns:
            Audit output as string.
        """
        args = ["audit", self.name]
        if security:
            args.append("--security")
        if json_output:
            args.append("--json")
        return self._run(args, capture=True)

    def status(self) -> str:
        """Show pod status and resource usage."""
        return self._run(["status", self.name], capture=True)

    def start(self) -> None:
        """Start the pod in background."""
        self._run(["start", self.name])

    def stop(self) -> None:
        """Stop the pod."""
        self._run(["stop", self.name])

    def lock(self) -> None:
        """Freeze the pod."""
        self._run(["lock", self.name])

    def unlock(self) -> None:
        """Resume a frozen pod."""
        self._run(["unlock", self.name])

    def restart(self) -> None:
        """Restart the pod (stop + start)."""
        self._run(["restart", self.name])

    def kill(self) -> None:
        """Terminate pod processes and rollback changes."""
        self._run(["kill", self.name])

    def destroy(self) -> None:
        """Remove the pod entirely."""
        try:
            self._run(["destroy", self.name])
        except PodError:
            pass  # Pod may already be destroyed

    def resize(self, cpus: Optional[float] = None, memory: Optional[str] = None,
               tmp_size: Optional[str] = None, max_pids: Optional[int] = None,
               gpu: Optional[bool] = None) -> None:
        """Resize pod resources (live if running, config if stopped)."""
        args = ["resize", self.name]
        if cpus is not None:
            args.extend(["--cpus", str(cpus)])
        if memory is not None:
            args.extend(["--memory", memory])
        if tmp_size is not None:
            args.extend(["--tmp-size", tmp_size])
        if max_pids is not None:
            args.extend(["--max-pids", str(max_pids)])
        if gpu is not None:
            args.extend(["--gpu", str(gpu).lower()])
        self._run(args)

    def vault_set(self, key: str, value: str) -> None:
        """Store a secret in the pod's vault."""
        self._run(["vault", self.name, "set", key, value])

    def init_with_base(self, config: Optional[str] = None, preset: Optional[str] = None,
                        base_name: Optional[str] = None, verbose: bool = False) -> None:
        """Create pod and save as base for fast cloning.

        Args:
            config: Path to pod.yaml config file.
            preset: Built-in preset name.
            base_name: Name for the base (defaults to pod name).
            verbose: Show live setup output.
        """
        args = ["init", self.name, "--create-base"]
        if base_name:
            args[-1] = f"--create-base={base_name}"
        cfg = config or self._config
        pre = preset or self._preset
        if cfg:
            args.extend(["-c", cfg])
        elif pre:
            args.extend(["--preset", pre])
        if verbose:
            args.append("--verbose")
        self._run(args)
        self._initialized = True

    @staticmethod
    def clone(source, name: str, mode: Optional[str] = None) -> 'Pod':
        """Clone a pod from a base (fast — ~8ms).

        Args:
            source: Source pod/base name (str) or Pod instance.
            name: Name for the new cloned pod.
            mode: Isolation mode (standard/full).

        Returns:
            A new Pod instance for the clone.
        """
        source_name = source.name if isinstance(source, Pod) else source
        pod = Pod(name, mode=mode)
        pod._run(["clone", source_name, name])
        pod._initialized = True
        return pod

    def detach(self) -> 'Pod':
        """Mark pod as persistent — won't auto-destroy on context exit.

        Returns self for chaining.
        """
        self._persist = True
        return self

    def start_display(self) -> Optional[str]:
        """Start the pod with web display (noVNC) and return the URL.

        Starts the pod in background with desktop environment.
        Returns the noVNC URL for browser access.
        """
        self.start()
        url = self.display_url
        if url:
            print(f"  Desktop → {url}")
        return url

    def snapshot_create(self, name: Optional[str] = None) -> None:
        """Create a snapshot of the pod's current overlay state."""
        args = ["snapshot", self.name, "create"]
        if name:
            args.extend(["--name", name])
        self._run(args)

    def snapshot_restore(self, name: str) -> None:
        """Restore a snapshot."""
        self._run(["snapshot", self.name, "restore", name])

    def snapshot_list(self) -> str:
        """List all snapshots."""
        return self._run(["snapshot", self.name, "ls"], capture=True)

    def snapshot_destroy(self, name: str) -> None:
        """Delete a snapshot."""
        self._run(["snapshot", self.name, "destroy", name])

    def logs(self) -> str:
        """Show pod output logs."""
        return self._run(["logs", self.name], capture=True)

    def info(self) -> dict:
        """Get pod info as a dict (name, status, IP, display URL, etc.)."""
        try:
            result = self._run(["ls", "--json"], capture=True)
            pods = json.loads(result)
            for p in pods:
                if p.get("name") == self.name:
                    return p
        except (PodError, json.JSONDecodeError):
            pass
        return {}

    @property
    def display_url(self) -> Optional[str]:
        """Get the noVNC display URL if web display is enabled."""
        info = self.info()
        return info.get("display_url") or info.get("novnc_url")

    @property
    def ip(self) -> Optional[str]:
        """Get the pod's IP address."""
        info = self.info()
        return info.get("ip") or info.get("pod_ip")

    @staticmethod
    def disposable(base, name: str, command: str,
                   commit_paths: Optional[List[str]] = None,
                   output: Optional[str] = None,
                   mode: Optional[str] = None,
                   root: bool = False,
                   env: Optional[dict] = None) -> Optional[str]:
        """Clone from base, run command, optionally save results, destroy.

        The fastest governed execution — ~8ms clone, run, commit, destroy.
        Equivalent to `docker run --rm` but with governance.

        Args:
            base: Base pod name to clone from.
            name: Name for the disposable pod.
            command: Shell command to run.
            commit_paths: Paths to commit (None = discard everything).
            output: Export committed files to this directory instead of host.
            mode: Isolation mode (standard/full).
            root: Run as root.
            env: Environment variables.

        Returns:
            Diff output as string if commit_paths is set, else None.
        """
        base_name = base.name if isinstance(base, Pod) else base
        pod = Pod.clone(base_name, name, mode=mode)
        try:
            pod.run(command, root=root, env=env)
            if commit_paths:
                diff = pod.diff()
                pod.commit(*commit_paths, output=output, rollback_rest=True)
                return diff
            return None
        finally:
            pod.destroy()

    @staticmethod
    def gc() -> None:
        """Clean up orphaned resources (iptables, cgroups, netns)."""
        binary = shutil.which("envpod")
        if binary:
            try:
                cmd = ["sudo", binary, "gc"] if _get_mode() == "full" else [binary, "gc"]
                subprocess.run(cmd, capture_output=True)
            except subprocess.CalledProcessError:
                pass

    def exists(self) -> bool:
        """Check if the pod already exists."""
        try:
            result = self._run(["ls", "--json"], capture=True)
            pods = json.loads(result)
            return any(p.get("name") == self.name for p in pods)
        except (PodError, json.JSONDecodeError):
            return False

    def _run(self, args: List[str], capture: bool = False) -> Optional[str]:
        """Execute an envpod command."""
        cmd = self._build_cmd(args)
        try:
            if capture:
                result = subprocess.run(
                    cmd, capture_output=True, text=True, check=True
                )
                return result.stdout
            else:
                subprocess.run(cmd, check=True)
                return None
        except subprocess.CalledProcessError as e:
            stderr = e.stderr or ""
            raise PodError(f"envpod {' '.join(args)} failed (exit {e.returncode}): {stderr.strip()}")

    def _build_cmd(self, args: List[str]) -> List[str]:
        """Build the full command with sudo if needed."""
        binary = shutil.which("envpod")
        if not binary:
            raise PodError("envpod binary not found. Install: curl -fsSL https://envpod.dev/install.sh | sudo bash")
        if self._mode == "full":
            return ["sudo", binary] + args
        return [binary] + args


def _get_mode() -> str:
    """Get the isolation mode from config or prompt interactively."""
    import os
    config_dir = os.path.expanduser("~/.config/envpod")
    config_file = os.path.join(config_dir, "sdk.json")

    # Check saved preference
    if os.path.exists(config_file):
        try:
            with open(config_file) as f:
                config = json.load(f)
                return config.get("mode", "full")
        except (json.JSONDecodeError, IOError):
            pass

    # Check environment variable
    env_mode = os.environ.get("ENVPOD_MODE")
    if env_mode in ("standard", "full"):
        return env_mode

    # Interactive prompt
    print()
    print("  envpod — choose isolation mode:")
    print("    [1] Standard — full governance, no sudo needed")
    print("        (no cgroup limits, no network namespace)")
    print("    [2] Full — complete isolation + governance (requires sudo)")
    print("        (cgroup limits, network namespace, DNS filtering)")
    print()

    try:
        choice = input("  Choice [1/2]: ").strip()
    except (EOFError, KeyboardInterrupt):
        choice = "2"

    mode = "standard" if choice == "1" else "full"

    # Save preference
    try:
        os.makedirs(config_dir, exist_ok=True)
        with open(config_file, "w") as f:
            json.dump({"mode": mode}, f)
        print(f"  Saved to {config_file}")
    except IOError:
        pass

    return mode
